Add optional HashType to RawData endpoint

AdamSLevy opened the pull request:
This allows the client to optionally specify the hash type when using the "raw-data" API method. For example, a user typically knows what the hash they are searching for refers to: an entry hash, an EBlock KeyMR, a FCT or EC TX hash. Now they can include that information and reduce needless database look ups. If none is specified, it falls back to searching all of the databases until it finds the hash. This path has also been streamlined.

AdamSLevy commented:
I haven't been able to get the tests passing on this but I'm not sure if that's because I don't have the right set up on my system.

PaulBernier commented on wsapi/wsapiV2.go:
```diff
@@ -722,40 +720,55 @@ func HandleV2RawData(state interfaces.IState, params interface{}) (interface{},
-		b, _ = block.MarshalBinary()
-	} else if block, _ = state.FetchFactoidTransactionByHash(h); block != nil {
-		b, _ = block.MarshalBinary()
-	} else if block, _ = state.FetchEntryByHash(h); block != nil {
```
you replaced the query `state.FetchEntryByHash(h)` by `dbase.FetchEntry(h)`, do you know the difference? Are you confident that doesn't change the behaviour?

AdamSLevy replied:
I do not know the difference and I am not confident it doesn't change the behavior. I hope someone more knowledgeable can vet that.

AdamSLevy replied:
@WhoSoup @carryforward

AdamSLevy replied:
Thanks for the input. Should both state.FetchEntryByHash and dbase.FetchEntry be called, or just state.FetchEntryByHash?

WhoSoup replied:
Just the state one, it calls dbase.FetchEntry already:
https://github.com/FactomProject/factomd/blob/c0f29f1452a03e36dd1115781ceab99ef2ec3cbf/state/ack.go#L611-L653

PaulBernier commented on wsapi/wsapiV2.go:
```diff
@@ -722,40 +720,55 @@ func HandleV2RawData(state interfaces.IState, params interface{}) (interface{},
+		}
+		if block, _ = dbase.FetchEntry(h); block != nil {
+			break
+		}
```
You mentioned that query for Entry is probably the most common so you could optimize the order by putting Entry first. Which I agree. Why did you still put tx queries first?

AdamSLevy replied:
When I mentioned that I forgot about tx hashes. I'm not sure which should take precedence: tx hashes or entry hashes. Perhaps entries should come before txs. But I think txs should come before the rest for some gut feeling reason??

PaulBernier replied:
Ok. I guess it depends on the kind of application, if it's one that work with the monetary side of Factom (like an exchange) or with the data side (pretty much all other uses cases). No big deal but I think I'd still put Entry first.

WhoSoup replied:
The amount of people using get-raw in the first place is pretty low, so if they are looking for an optimization they can just specify the hash type.

AdamSLevy commented:
> It's a neat addition.
> Can you add 1 or 2 simple test in `wsapiV2_test` that uses your new parameter? There is already a `TestHandleV2GetRaw` there, should be easy enough.

Will do! I'll take care of that tomorrow. I couldn't get the existing tests to pass on my system. I didn't have a database built though...

WhoSoup commented on wsapi/wsapiV2.go:
```diff
@@ -722,40 +720,55 @@ func HandleV2RawData(state interfaces.IState, params interface{}) (interface{},
+		block, _ = dbase.FetchABlock(h)
+	case HashTypeDBlock:
+		block, _ = dbase.FetchDBlock(h)
+	default:
```
Technically, changing the order in which blocks are searched is a change in functionality though with SHA256 I don't think the likelihood of a conflict existing in the first place is even there. This change should be fine unless I overlooked something obvious.

WhoSoup commented on wsapi/wsapiV2.go:
```diff
@@ -722,40 +720,55 @@ func HandleV2RawData(state interfaces.IState, params interface{}) (interface{},
+		if block, _ = state.FetchECTransactionByHash(h); block != nil {
+			break
+		}
+		if block, _ = dbase.FetchEntry(h); block != nil {
```
Same as the other comment, this should be the state lookup function.

AdamSLevy replied:
Does it need both?

WhoSoup replied:
I think so. Writing entries to the database and the processlist are asynchronous, so you might run into race conditions where you have knowledge of the hash but it's not written to the database yet. I'm not 100% certain it's possible, you'd have to do a deep dive of factomd functionality to double check.

AdamSLevy replied:
doesn't hurt to use both I suppose. I guess both should be used under the HashTypeEntry case as well then...

AdamSLevy replied:
but in another comment you said that state.FetchEntryByHash calls dbase.FetchEntry()

WhoSoup replied:
> but in another comment you said that state.FetchEntryByHash calls dbase.FetchEntry()

Yes, it's the third and last place to look for an entry. It tries to get it from the processlist and holding map first.

WhoSoup commented on wsapi/wsapiV2.go:
```diff
@@ -713,7 +712,6 @@ func HandleV2RawData(state interfaces.IState, params interface{}) (interface{},
 	hashkey := new(HashRequest)
 	err := MapToObject(params, hashkey)
 	if err != nil {
-		panic(reflect.TypeOf(params))
```
Technically, this is a change in functionality, though it only affects API requests that send a non-string type for the "Hash" parameter. The change should be fine under the assumption that current clients send valid parameters.

AdamSLevy replied:
Is panicking really the right response to a bad type? Seems like it should return an invalid params. Is that what panicking does?

If panicking is the appropriate thing, the return statement should be removed as its dead code, which was confusing to me as it was not obvious that the panic wasn't just a remnant of someone's debugging process at one point.

WhoSoup replied:
No, it's not. It's definitely leftover debug code that ideally shouldn't have been there. It's not a question of right vs wrong, but "changing functionality". Removing bugs can sometimes break programs that depend on those bugs to work.

**I don't think this is one of those cases and the change is fine.**



> Is panicking really the right response to a bad type? Seems like it should return an invalid params. Is that what panicking does?

It's not. panicking will kill the http thread and produce the panic in the console (without impacting the node at all). The client will get no output at all.

I tested it just to make sure and a bad request on the server side looks like this:
```
2020/02/13 08:32:03 http: panic serving x.x.x:57191: map[string]interface {}
goroutine 781658 [running]:
net/http.(*conn).serve.func1(0xc0032d0640)
        /usr/local/go/src/net/http/server.go:1769 +0x139
panic(0xdc2920, 0xccf6e0)
        /usr/local/go/src/runtime/panic.go:522 +0x1b5
github.com/FactomProject/factomd/wsapi.HandleV2RawData(0x1089640, 0xc00016e000, 0xccf6e0, 0xc00b24fce0, 0x0, 0x0, 0x0)
        /go/src/github.com/FactomProject/factomd/wsapi/wsapiV2.go:677 +0x92d
github.com/FactomProject/factomd/wsapi.HandleV2JSONRequest(0x1089640, 0xc00016e000, 0xc002ea49c0, 0x0, 0x0)
        /go/src/github.com/FactomProject/factomd/wsapi/wsapiV2.go:131 +0x676
github.com/FactomProject/factomd/wsapi.HandleV2(0x10644a0, 0xc0005ac1c0, 0xc023c47300)
        /go/src/github.com/FactomProject/factomd/wsapi/wsapiV2.go:65 +0x3dc
github.com/FactomProject/factomd/wsapi.APILogger.func1.1(0x10644a0, 0xc0005ac1c0, 0xc023c47300)
        /go/src/github.com/FactomProject/factomd/wsapi/server.go:96 +0x16d
github.com/FactomProject/factomd/wsapi.IDInjector.func1.1(0x10644a0, 0xc0005ac1c0, 0xc023c47300)
        /go/src/github.com/FactomProject/factomd/wsapi/server.go:106 +0x136
net/http.HandlerFunc.ServeHTTP(0xc00226c3e0, 0x10644a0, 0xc0005ac1c0, 0xc023c47300)
        /usr/local/go/src/net/http/server.go:1995 +0x44
github.com/FactomProject/factomd/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc001c940c0, 0x10644a0, 0xc0005ac1c0, 0xc023c47100)
        /go/src/github.com/FactomProject/factomd/vendor/github.com/gorilla/mux/mux.go:210 +0xe3
net/http.serverHandler.ServeHTTP(0xc003f604e0, 0x10644a0, 0xc0005ac1c0, 0xc023c47100)
        /usr/local/go/src/net/http/server.go:2774 +0xa8
net/http.(*conn).serve(0xc0032d0640, 0x1067760, 0xc002ea4880)
        /usr/local/go/src/net/http/server.go:1878 +0x851
created by net/http.(*Server).Serve
        /usr/local/go/src/net/http/server.go:2884 +0x2f4
```

and on the client side:
```
curl --data-binary "{\"jsonrpc\": \"2.0\", \"id\": 0, \"method\": \"raw-data\", \"params\": {\"hash\": 1}}" -H "content-type:text/plain;" http://x.x.x:8088/v2
curl: (52) Empty reply from server
```

AdamSLevy replied:
Thanks for the thorough response.

AdamSLevy commented:
I just fixed the state vs dbase lookup for Entry Hashes and I added tests where the HashType is specified. 

I can't get the tests to pass probably due to not having the right test database. Maybe someone who has these tests passing can confirm they work. Maybe the CI does that...
